### ISSUE TITLE
Add cancellation support to twisted.internet.defer.DeferredList

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6639]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| kaizhang reported|
|-|-|
|Trac ID|trac#6639|
|Type|enhancement|
|Created|2013-07-22 09:14:24Z|
|Branch|https://github.com/twisted/twisted/tree/deferredlist-cancellation-6639|

Add cancellation support to `twisted.internet.defer.DeferredList`. When a `DeferredList` is cancelled by calling the `cancel` method, cancel every `Deferred` in the list. If the `DeferredList` has already fired, including the case where the `fireOnOneCallback`/`fireOnOneErrback` flag is set and the `DeferredList` fires because one `Deferred` in the list fires with a non-failure/failure result, do nothing in the `cancel` method.

<details><summary>Searchable metadata</summary>

```
trac-id__6639 6639
type__enhancement enhancement
reporter__kaizhang kaizhang
priority__normal normal
milestone__ 
branch__branches_deferredlist_cancellation_6639 branches/deferredlist-cancellation-6639
branch_author__kaizhang kaizhang
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1374484464000000 1374484464000000
changetime__1377065613000000 1377065613000000
version__None None
owner__kaizhang kaizhang

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|kaizhang commented|
|-|-|

(In [39276]) Branching to 'deferredlist-cancellation-6639'

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|kaizhang commented|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deferredlist-cancellation-6639)

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|kaizhang set owner to kaizhang|
|-|-|

Change the `deferredList` instance variable to private.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kaizhang set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|kaizhang removed owner<br>kaizhang set status to `new`|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deferredlist-cancellation-6639)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit set owner to @habnabit|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_7"></a>|@habnabit set owner to kaizhang|
|-|-|

1.This breaks backward compatibility by iterating over `deferredList` twice.
    There is likely to be code that is passing `DeferredList` a generator or
    some other single-use iterator. The easy fix is to iterate over
    `self._deferredList`.

 2. The docstring for `gatherResults` doesn't seem correct. What's guaranteeing
    that the returned `Deferred` will errback? All of the passed-in `Deferred`s
    can callback instead of errback on cancellation.

 3. Similar to #2 for the `DeferredList.cancel` docstring.

 4. Similar to #2, there don't seem to be any tests for the case where
    cancelling a Deferred callbacks instead of errbacks for `DeferredList` or
    `gatherResults`.

 5. Several places in test_defer.py say L{defer.Deferrd}.

 6. You might consider making a function or method to instantiate a deferred
    and set its cancel method to `MockCancel`, considering this is a thing you
    do a lot.

Thank you for your contribution!

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

I'd also add the `MockCall` is completely unnecessary, you don't need to mock anything to test that a Deferred has been cancelled.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Thanks for your work on this so far kaizhang, habnabit, itamar.

First, I agree with Itamar.  I'd suggest getting rid of `MockCancel` entirely.  It's better to test the actual effects of cancellation.  You can assert that the Deferreds have been cancelled by asserting that they have a `CancelledError` failure result.

Also, think about not logging any errors at all here.  As far as I can tell, the only exception that might get logged is `AlreadyCalledError` - which actually just reflects a shortcoming of the implementation of cancellation here.  If `DeferredList.cancel` didn't try to cancel already-called Deferreds, then it wouldn't provoke this exception.  If there's another case, please demonstrate it with a unit test (using `flushLoggedErrors(Exception)` makes the test less clear than it could be about what's going to happen - every possible logged error is an `Exception`, so who knows what this will flush?).

Also, try writing unit tests that only have a single assertion in them.  These are more useful, since you get to see the results of all assertions in a single test run.  When you have two assertions, the first can fail and prevent the second from running.  Then you can't tell if the second would have passed or failed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

`cancel()` never throws an exception, not even `AlreadyCalledError` (except for something that is arguably a bug in the implementation and we should probably fix, exceptions from cancellation functions).

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|kaizhang commented|
|-|-|

Replying to [itamar](#note_8):
> I'd also add the `MockCall` is completely unnecessary, you don't need to mock anything to test that a Deferred has been cancelled.

The `MockCancel` is not used to test that a `Deferred` has been cancelled. It's used to test that when the `DeferredList` has already fired, the `Deferred`s in the list will NOT be cancelled. Is there anyway else to test this?

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|kaizhang commented|
|-|-|

Replying to [exarkun](#note_9):
> Also, think about not logging any errors at all here.  As far as I can tell, the only exception that might get logged is `AlreadyCalledError` - which actually just reflects a shortcoming of the implementation of cancellation here.  If `DeferredList.cancel` didn't try to cancel already-called Deferreds, then it wouldn't provoke this exception.  If there's another case, please demonstrate it with a unit test (using `flushLoggedErrors(Exception)` makes the test less clear than it could be about what's going to happen - every possible logged error is an `Exception`, so who knows what this will flush?).

I was worried about the user might supply a canceller which raise a exception accidentally.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

> I was worried about the user might supply a canceller which raise a exception accidentally.

Okay.  That seems ... a bit more reasonable.  Please change the `log.err` call to pass a reason argument explaining that this is the case.  It's sort of a historical accident and a misfeature that you can call `log.err` without a reason - this produces unexplained errors in the log that are often very difficult to track down.

> The `MockCancel` is not used to test that a `Deferred` has been cancelled. It's used to test that when the `DeferredList` has already fired, the `Deferred`s in the list will NOT be cancelled. Is there anyway else to test this?

You can use `TestCase.assertNoResult` to verify that a `Deferred` has not been fired (which also means it has not been cancelled).

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|kaizhang commented|
|-|-|

Replying to [exarkun](#note_13):
> > The `MockCancel` is not used to test that a `Deferred` has been cancelled. It's used to test that when the `DeferredList` has already fired, the `Deferred`s in the list will NOT be cancelled. Is there anyway else to test this?
> 
> You can use `TestCase.assertNoResult` to verify that a `Deferred` has not been fired (which also means it has not been cancelled).
I used `TestCase.assertNoResult` in `test_cancelFiredOnOneCallbackDeferredList` to verify the `Deferred` has not been fired.

However, in `test_cancelDeferredListFiredWithNonFailure` and `test_cancelDeferredListFiredWithFailure` all the `Deferred`s have already fired(thus make the `DeferredList` fire), so I can not use `TestCase.assertNoResult` to verify the `DeferredList.cancel` didn't try to cancel all the `Deferred`s(after they fired). That's why I used the `MockCancel`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

> However, in test_cancelDeferredListFiredWithNonFailure and test_cancelDeferredListFiredWithFailure all the Deferreds have already fired(thus make the DeferredList fire), so I can not use TestCase.assertNoResult to verify the DeferredList.cancel didn't try to cancel all the Deferreds(after they fired). That's why I used the MockCancel.

What's the point of this test?  Or put another way, I don't think that it is an important feature of `DeferredList.cancel` that it not cancel its element `Deferred`s when they all have results already.  Is there any observable difference in behavior if you don't provide this guarantee?

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|kaizhang commented|
|-|-|

Replying to [exarkun](#note_15):
> What's the point of this test?  Or put another way, I don't think that it is an important feature of `DeferredList.cancel` that it not cancel its element `Deferred`s when they all have results already.  Is there any observable difference in behavior if you don't provide this guarantee?
There is no observable difference since `Deferred.cancel` itself won't try to cancel an already fired `Deferred`. This test is more about testing the logic is right(do not cancel an already fired `DeferredList`). So is it OK to just remove this test?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

I'd say it's okay to remove both the test *and* the logic. :)  There's no sense having code that doesn't produce observable differences.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|kaizhang commented|
|-|-|

Replying to [exarkun](#note_17):
> I'd say it's okay to remove both the test *and* the logic. :)  There's no sense having code that doesn't produce observable differences.

I think we can remove the test but keep the logic. The logic doesn't only cover the case where the `DeferredList` fires because all the `Deferred`s in the list has fired. It also covers the case where the `DeferredList` fires because one `Deferred` in the list has fired but the rest of the `Deferred`s may haven't fired(when the `fireOnOneCallback`/`fireOnOneErrback` flag is set). In this case, cancel the `Deferred` directly will case a problem.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

> I think we can remove the test but keep the logic.

This is always the wrong conclusion. :)  Untested code is not allowed in Twisted.

>  It also covers the case where the DeferredList fires because one Deferred in the list has fired but the rest of the Deferreds may haven't fired(when the fireOnOneCallback/fireOnOneErrback flag is set). In this case, cancel the Deferred directly will case a problem.

Great!  It sounds like there is some observable behavior after all.  Perhaps you can write a test for that case to replace the test we've been discussing?

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|kaizhang commented|
|-|-|

Replying to [exarkun](#note_19):
> Great!  It sounds like there is some observable behavior after all.  Perhaps you can write a test for that case to replace the test we've been discussing?

We do have the tests for that case, the `test_cancelFiredOnOneCallbackDeferredList` and `test_cancelFiredOnOneErrbackDeferredList`. So can we keep the logic as long as we keep these two tests?

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|kaizhang removed owner|
|-|-|

I have revised my code according to the comments and also added the news file.

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deferredlist-cancellation-6639)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_22"></a>|@habnabit commented|
|-|-|

Just a quick note: you did not address one of my review points:

    There don't seem to be any tests for the case where cancelling a Deferred callbacks instead of errbacks for `DeferredList` or `gatherResults`.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kaizhang set owner to kaizhang<br>kaizhang set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_24"></a>|kaizhang commented|
|-|-|

Replying to [habnabit](#note_22):
> Just a quick note: you did not address one of my review points:
> 
>     There don't seem to be any tests for the case where cancelling a Deferred callbacks instead of errbacks for `DeferredList` or `gatherResults`.

I will address this soon.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_25"></a>|kaizhang removed owner<br>kaizhang set status to `new`|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deferredlist-cancellation-6639)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_26"></a>|@itamarst set owner to kaizhang|
|-|-|

Thanks for all you work.

 1. News file should also mention `gatherResults`.
 1. In `test_cancelDeferredListWithException` you raise `Exception`, and then assert later on that the one `Exception` was logged. This is a bit of a problem, since you might have a bug in the code causing `AttributeError` or `NameError`. Since these are subclasses of `Exception`, the `flushLoggedErrors` will still catch those I believe, erroneously making the test pass. It's better to therefore use a subclass of `Exception` that is unlikely to be raised by the actual code, e.g. `RuntimeError`.
 1. `test_cancelDeferredListWithFireOnOneErrback` should assert the `FirstError` wraps a `CancelledError`.
 1. Same for `test_cancelGatherResults`.

Please address these comments, and then merge.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_27"></a>|kaizhang set status to `closed`|
|-|-|

(In [39657]) Merge deferredlist-cancellation-6639: Added cancel() method to DeferredList.

Author: kaizhang
Reviewer: itamar
Fixes: [#6639](https://github.com/twisted/twisted/issues/6639)

Added cancellation support to twisted.internet.defer.DeferredList. When a DeferredList is cancelled by calling the cancel method, cancel every Deferred in the list. If the DeferredList has already fired, including the case where the fireOnOneCallback/fireOnOneErrback flag is set and the DeferredList fires because one Deferred in the list fires with a non-failure/failure result, do nothing in the cancel method.
